### DISCUSSION TITLE
How to show .png file and use the SD Card

### DISCUSSION THREAD

**Discussion by wilwizard1979** [HASwitchPlate/openHASP#459] [Q&A] [answered]
Hi All,

I´m unable to add a .png to my display. It shows like this (white box, "No Data"):
![image](https://user-images.githubusercontent.com/108979356/229344944-b5e3e2e9-a3cf-48e9-9e9a-018ecc9b49db.png)

The syntax I´m using is the following (supressed):
```
{"page":4,"comment":" ---- Weather ----"}
{"page":4,"id":1,"obj":"obj","action":"none","x":0,"y":0,"w":480,"h":320,"bg_color":"#000000","radius":0,"border_side":0}
{"page":4,"id":2,"obj":"obj","x":4,"y":55,"w":230,"h":105,"bg_color":"#000000","border_color":"#000000","click":0}
{"page":4,"id":21,"obj":"img","src":"L:/testpng.png","parentid":2,"auto_size":1,"w":128,"offset_x":-6,"offset_y":-10}
{"page":4,"id":22,"obj":"label","x":100,"y":5,"w":130,"h":25,"align":"center","text":"date current","parentid":2}
{"page":4,"id":23,"obj":"label","x":125,"y":30,"w":95,"h":40,"align":"center","text":"00.0°C","parentid":2,"text_font":32}
.......
```

This is how the files are stored:
![image](https://user-images.githubusercontent.com/108979356/229345023-7fa7c0f0-bacb-45c1-9794-c026eb293461.png)

Second question, I have an SD Card installed and could not find a way to store the files on it. I just see the "L:" drive. Is there any special preparation (eg.: formatting) needed?

**Comment by adorobis** [comment, marked as answer]
Check below comments from [documentation](https://www.openhasp.com/0.6.3/design/objects/#image). Your image might be too large, you might not have the PSram or the png is not in a correct format:

_Either PNG or BIN image files are supported, from flash:_
_PNG image are decoded and kept in memory, thus you'll be able to display full-screen images only if your microcontroller [has installed PSram memory](https://www.openhasp.com/0.6.3/getting-started/#recommended-boards), else you will be limited to small icons or BIN files._
_BIN images are not stored in memory and are read from flash each time. This will be slower but allows the use of larger images even if your microcontroller doesn't have PSram._
_BIN images from flash do not support zoom or angle transformations._
_Use the online [LVGL Image Converter](https://lvgl.io/tools/imageconverter) to create BIN files. Set the output format to Binary or Binary RGB565._

**Reply by stovedoctor** [reply]
More info?
Type of board?
Flash and psram memory size?
Can you view\open the image when you click on it in the file editor file directory?

**Comment by wilwizard1979** [comment]
Hi, thanks for your replies. I needed some time to check the hints you provided.

Hi @adorobis, I just tested the LVL converted and created the bin file as you suggested. It works! But as you know, handling bin files is very limited specially in relation to autosize functions. So I would like to check whether .png files are possible to be read.

The Board is a Sunton ESP32-3248S035C (with Capacitive Touch). Link in [AliExpress Sunton Board](https://pt.aliexpress.com/item/1005004632953455.html?spm=a2g0o.order_list.order_list_main.52.2dd8caa4tO6zxs&gatewayAdapt=glo2bra).

According to the documentation, which is not specific to this capacitive model, there is a 520KB SRAM and 4MB Flash.
I tried different .png images, including the one part of the Openhasp Image Function description [Image Example](https://www.openhasp.com/0.6.3/assets/images/objects/lv_ex_img_1.png).

Unfortunately the result is the same as shown above.
One thing I noticed is that, if I refer for a file that does not exist (for testing purposes) it shows a blank image. But when I refer to any of the images I´m testing, I see a white square written "No data" (as shown above). So, it shows that the file is found but for some reason it cannot identify.

**Reply by fvanroie** [reply]
The 3.5" Sunton model does not have PSram. PNG images are decoded and stored into RAM which is not possible with just 520kB of sRam.
So you need a Sunton model with ESP32-S3 (4.3" or larger) to use PNG images in PSram...

**Reply by wilwizard1979** [reply]
> The 3.5" Sunton model does not have PSram. PNG images are decoded and stored into RAM which is not possible with just 520kB of sRam.
> So you need a Sunton model with ESP32-S3 (4.3" or larger) to use PNG images in PSram...

Thanks a lot for the clarification. I will consider buying a larger version then, but investment vs. benefit in comparison to an Amazon Fire tablet (in Prime Day) starts to not being sustainable.

Is the SD card slot then useless in this case?

**Reply by fvanroie** [reply]
SD card is not supported in openHASP at this time.

**Reply by wilwizard1979** [reply]
Thanks @fvanroie. Do you think that when this will be ever possible, images could be directly read without the need of PSRAM?

**Reply by fvanroie** [reply]
You can already use BIN files from flash without PSram. So I would assume this can also work from SD card if you implement a driver for it.

**Reply by wilwizard1979** [reply]
any hint on how I could do this?
I´m not very experienced on developing drivers, but I would be brave enough to learn and try.

**Reply by elik745i** [reply]
Any update to above project? I also got no data when I use large png, however if I resize it to lıke 40x20pixels it works on ESP32. Any progress implementing sd card support?

**Reply by ouchkilljoys** [reply]
Yeah would be cool to see SD card support for this!
